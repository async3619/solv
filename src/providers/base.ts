import fetch from "node-fetch";

import { Challenge, InputType } from "../utils/types";
import { CachedChallenge } from "../utils/retrieveChallenge";

export abstract class BaseProvider {
    public abstract checkUrl(url: string): boolean;
    public abstract retrieve(url: string): Promise<Challenge>;
    public abstract getId(url: string): string | number;

    protected constructor(public readonly needJs: boolean) {}

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public hydrateCache({ version, ...challenge }: CachedChallenge): Challenge {
        return {
            ...challenge,
            provider: this,
        };
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem !== "string" ? JSON.stringify(outputItem) : outputItem;
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public beforeExecute(sourceCode: string, input: InputType): string {
        return sourceCode;
    }

    public getName(): string {
        return this.constructor.name.replace("Provider", "");
    }

    protected async fetch(url: string): Promise<string> {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch (${response.status}) - ${response.statusText}`);
        }

        return response.text();
    }
}
