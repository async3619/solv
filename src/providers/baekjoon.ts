import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";
import { Challenge } from "../types";

export class BaekjoonProvider extends BaseProvider {
    private readonly urlRegex = /^https?:\/\/(www\.)?acmicpc\.net\/problem\/([0-9]*?)$/;

    public checkUrl(url: string) {
        return this.urlRegex.test(url);
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        const htmlCode = await this.fetch(url);
        const document = parseDOM(htmlCode);

        const titleDOM = document.querySelector("#problem_title");
        const descriptionDOM = document.querySelector("#problem_description");
        const inputDescriptionDOM = document.querySelector("#problem_input");
        const outputDescriptionDOM = document.querySelector("#problem_output");
        const ioItem = document.querySelectorAll(".sampledata");

        const ioText: [boolean, string][] = [];
        for (const item of ioItem) {
            ioText.push([item.id.includes("-input-"), item.innerHTML.trim()]);
        }

        if (
            !titleDOM?.textContent ||
            !descriptionDOM?.textContent ||
            !inputDescriptionDOM?.textContent ||
            !outputDescriptionDOM?.textContent
        ) {
            throw new Error("Failed to parse challenge information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            inputDescription: inputDescriptionDOM.textContent.trim(),
            outputDescription: outputDescriptionDOM.textContent.trim(),
            input: ioText.filter(([i]) => i).map(([, t]) => t),
            output: ioText.filter(([i]) => !i).map(([, t]) => t.replace(/\r\n/g, "\n")),
            provider: this,
            initialCode: `
function solution(input: string[]) {
    // ...
}

(() => {
    const readline = require("readline");

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const input: string[] = [];
    rl.on("line", line => {
        input.push(line);
    }).on("close", function () {
        solution(input);
        process.exit();
    });
})();
`.trim(),
        };
    }
}
