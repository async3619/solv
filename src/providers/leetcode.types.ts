export interface QuestionMetadata {
    name: string;
    params: Array<{
        name: string;
        type: string;
    }>;
    return: {
        type: string;
    };
}
