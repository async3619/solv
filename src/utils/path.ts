import * as os from "os";
import * as path from "path";
import * as fs from "fs-extra";
import { v4 as generateUUID } from "uuid";

const INSTANCE_ID = generateUUID();

export async function getTranspiledPath() {
    const targetDirectory = path.join(os.homedir(), ".solv");

    await fs.ensureDir(targetDirectory);
    return path.join(targetDirectory, `${INSTANCE_ID}.js`);
}
