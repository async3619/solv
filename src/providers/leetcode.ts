import { assertParse, is } from "typia";
import { parse as parseDOM } from "node-html-parser";
import { Client, CombinedError, createClient } from "@urql/core";
import { stripHtml } from "string-strip-html";

import { BaseProvider } from "./base";

import { Challenge, InputType } from "../utils/types";
import { LeetCodeQuestionDocument } from "../queries/leetcode";
import type { LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables } from "../queries.data";
import { QuestionMetadata } from "./leetcode.types";

const URL_CHECK_REGEX = /(https?:\/\/)?leetcode\.com\/problems\/([a-z0-9-]+)\/?$/;

export class LeetCodeProvider extends BaseProvider {
    private readonly client: Client;

    public constructor() {
        super(false);

        this.client = createClient({
            url: "https://leetcode.com/graphql",
            requestPolicy: "network-only",
            fetchOptions: () => {
                return {
                    method: "GET",
                    headers: {
                        "content-type": "application/json",
                    },
                };
            },
        });
    }

    public checkUrl(url: string): boolean {
        return URL_CHECK_REGEX.test(url);
    }

    public getId(url: string): string | number {
        const match = URL_CHECK_REGEX.exec(url);
        if (!match) {
            throw new Error("Failed to parse challenge ID.");
        }

        return match[2];
    }

    public beforeExecute(sourceCode: string, input: InputType, challenge: Challenge): string {
        let entryPointName = "solution";
        if (challenge.initialCode) {
            const match = challenge.initialCode.match(/function ([a-zA-Z0-9]+)\(/);
            if (match) {
                entryPointName = match[1];
            }
        }

        return [sourceCode, `console.info(${entryPointName}(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const titleSlug = this.getId(url);
        if (!is<string>(titleSlug)) {
            throw new Error("Failed to parse challenge ID.");
        }

        try {
            const { error, data } = await this.client
                .query<LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables>(LeetCodeQuestionDocument, { titleSlug })
                .toPromise();

            if (error) {
                throw error;
            }

            if (!data?.question) {
                throw new Error("Failed to retrieve challenge data.");
            }

            const codeSnippet = data.question.codeSnippets.find(item => item.langSlug === "typescript");
            if (!codeSnippet) {
                throw new Error("This challenge does not support TypeScript.");
            }

            const document = parseDOM(data.question.content);
            const testCases = [...document.querySelectorAll("pre")].filter(item => {
                if (!item.parentNode) {
                    return false;
                }

                if (!item.previousElementSibling) {
                    return false;
                }

                const title = item.previousElementSibling.querySelector("strong.example");
                if (!title) {
                    return false;
                }

                return true;
            });

            const metadata = assertParse<QuestionMetadata>(data.question.metaData);
            const testCaseSummaries = testCases.map(item => stripHtml(item.textContent.trim()).result);
            const inputs: InputType[] = [];
            const outputs: string[] = [];
            for (const summary of testCaseSummaries) {
                const [input, output] = summary.split("\n").map(item => {
                    return item.trim().replace(/^[A-Za-z]*?: /, "");
                });

                const inputParameters = input.split(/, /).map(item => item.split(" = "));
                const data: InputType = [];
                for (const [key, value] of inputParameters) {
                    const index = metadata.params.findIndex(item => item.name === key);
                    if (index === -1) {
                        throw new Error(`Failed to find parameter "${key}".`);
                    }

                    data[index] = value;
                }

                inputs.push(data);
                outputs.push(output.replace("Output: ", ""));
            }

            return {
                id: data.question.questionId,
                title: data.question.questionTitle,
                description: stripHtml(data.question.content).result,
                provider: this,
                initialCode: codeSnippet.code,
                input: inputs,
                output: outputs,
            };
        } catch (e) {
            if (e instanceof CombinedError) {
                if (e.networkError) {
                    throw e.networkError;
                } else if (e.graphQLErrors && e.graphQLErrors.length > 0) {
                    throw e.graphQLErrors[0];
                }
            }

            throw e;
        }
    }
}
