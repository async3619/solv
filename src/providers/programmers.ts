import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";

import { findSiblingElement } from "../utils/findSiblingElement";
import { parseSampleTable } from "../utils/parseSampleTable";
import { Challenge } from "../utils/types";

export class ProgrammersProvider extends BaseProvider {
    public constructor() {
        super(false);
    }

    public checkUrl(url: string): boolean {
        return /(https?:\/\/)?(school\.)?programmers\.co\.kr\/learn\/courses\/([0-9]*?)\/lessons\/[0-9]([0-9]*?)$/.test(
            url,
        );
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem === "string" ? `"${outputItem}"` : JSON.stringify(outputItem);
    }

    public beforeExecute(sourceCode: string): string {
        return [sourceCode, `console.info(solution(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        targetUrl.searchParams.set("language", "javascript");

        const htmlCode = await this.fetch(targetUrl.toString());

        const document = parseDOM(htmlCode);
        const descriptionDOM = document.querySelector("#tour2 > div > div");
        const titleDOM = document.querySelector("#tab > li");
        if (!titleDOM?.textContent || !descriptionDOM?.textContent) {
            throw new Error("Failed to parse challenge information.");
        }

        const headers = document.querySelectorAll(".guide-section h5");
        const ioSampleHeaderRegex = /^입출력 ?예$/;
        const ioSampleHeaderDOM = headers.find(header => ioSampleHeaderRegex.test(header.textContent.trim()));
        if (!ioSampleHeaderDOM) {
            throw new Error("Failed to parse challenge input / output samples.");
        }

        const ioSampleTable = findSiblingElement(ioSampleHeaderDOM, node => node.tagName === "TABLE");
        if (!ioSampleTable) {
            throw new Error("Failed to find challenge output sample table.");
        }

        const initialCode = document
            .querySelector(`input[type="hidden"][name^="initial_code_"]`)
            ?.getAttribute("value");

        if (!initialCode) {
            throw new Error("Failed to parse initial code information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            ...parseSampleTable(ioSampleTable),
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            provider: this,
            initialCode,
        };
    }
}
