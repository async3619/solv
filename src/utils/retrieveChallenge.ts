import * as path from "path";
import * as os from "os";
import * as fs from "fs-extra";
import * as yaml from "yaml";

import { BaseProvider } from "../providers/base";
import { Challenge } from "./types";
import { getVersion } from "./cli";

const CACHE_DIR = path.join(os.homedir(), ".solv", "caches");

export type CachedChallenge = Omit<Challenge, "provider"> & { version: string };

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function stripFields({ provider, ...rest }: Challenge) {
    return { ...rest };
}

export async function retrieveChallenge(
    url: string,
    provider: BaseProvider,
    noCache = false,
    cacheDir: string = CACHE_DIR,
): Promise<Challenge> {
    if (noCache) {
        return provider.retrieve(url);
    } else {
        const targetPath = path.join(CACHE_DIR, `${provider.getName().toLowerCase()}_${provider.getId(url)}.yml`);
        await fs.ensureDir(cacheDir);

        if (!fs.existsSync(targetPath)) {
            const challenge = await provider.retrieve(url);
            const cachedChallenge: CachedChallenge = {
                ...stripFields(challenge),
                version: getVersion(),
            };

            await fs.writeFile(targetPath, yaml.stringify(cachedChallenge));
            return challenge;
        } else {
            const content = await fs.readFile(targetPath).then(buf => buf.toString());
            const challenge: CachedChallenge = yaml.parse(content);
            if (challenge.version !== getVersion()) {
                await fs.unlink(targetPath);
                return retrieveChallenge(url, provider, noCache, cacheDir);
            }

            return provider.hydrateCache(challenge);
        }
    }
}
