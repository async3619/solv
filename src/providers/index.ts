import { BaseProvider } from "./base";

import { BaekjoonProvider } from "./baekjoon";
import { ProgrammersProvider } from "./programmers";
import { LeetCodeProvider } from "./leetcode";

const TARGET_PROVIDERS: BaseProvider[] = [new BaekjoonProvider(), new ProgrammersProvider(), new LeetCodeProvider()];

export function getProvider(targetUrl: string) {
    const provider = TARGET_PROVIDERS.find(provider => provider.checkUrl(targetUrl));
    if (!provider) {
        throw new Error("there was no matched coding challenge service provider on records. abort.");
    }

    return provider;
}
