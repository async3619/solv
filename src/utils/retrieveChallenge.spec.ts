import * as fs from "fs-extra";
import * as yaml from "yaml";
import * as path from "path";

import { CachedChallenge, retrieveChallenge } from "./retrieveChallenge";

import { BaekjoonProvider } from "../providers/baekjoon";

describe("retrieveChallenge", () => {
    it("should retrieves challenge information correctly", async () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { provider, ...data } = await retrieveChallenge(
            "https://www.acmicpc.net/problem/1018",
            new BaekjoonProvider(),
            true,
        );

        expect(data).toMatchSnapshot();
    });

    it("should creates cache for challenge information", async () => {
        jest.mock("fs-extra");

        let saved = false;
        (fs.writeFile as any) = async () => {
            saved = true;
        };
        (fs.ensureDir as any) = () => Promise.resolve();
        (fs.existsSync as any) = () => false;

        await retrieveChallenge("https://www.acmicpc.net/problem/1018", new BaekjoonProvider());
        expect(saved).toBeTruthy();
    });

    it("should reads cache if there was a valid cache file", async () => {
        jest.mock("fs-extra");

        let read = false;
        const readFile = fs.readFile;
        (fs.readFile as any) = () => {
            read = true;
            return readFile(path.join(__dirname, "__snapshots__", "cache.yml"));
        };
        (fs.ensureDir as any) = () => Promise.resolve();
        (fs.existsSync as any) = () => true;

        const data = await retrieveChallenge("https://www.acmicpc.net/problem/1018", new BaekjoonProvider());

        expect(read).toBeTruthy();
        expect(data.provider.getName()).toBe("Baekjoon");

        (fs.readFile as any) = readFile;
    });

    it("should invalidates cache if cache version is not matched", async () => {
        jest.mock("fs-extra");

        let unlinked = false;
        const readFile = fs.readFile;
        (fs.readFile as any) = async () => {
            const data = await readFile(path.join(__dirname, "__snapshots__", "cache.yml")).then(buf => buf.toString());
            const cache: CachedChallenge = yaml.parse(data);
            cache.version = "0.0.0";

            return Buffer.from(yaml.stringify(cache));
        };
        (fs.ensureDir as any) = () => Promise.resolve();
        (fs.unlink as any) = () => (unlinked = true);
        (fs.existsSync as any) = () => !unlinked;

        await retrieveChallenge("https://www.acmicpc.net/problem/1018", new BaekjoonProvider());
        expect(unlinked).toBeTruthy();
    });
});
