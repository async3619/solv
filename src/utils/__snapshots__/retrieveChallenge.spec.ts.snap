// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retrieveChallenge should retrieves challenge information correctly 1`] = `
{
  "description": "지민이는 자신의 저택에서 MN개의 단위 정사각형으로 나누어져 있는 M×N 크기의 보드를 찾았다. 어떤 정사각형은 검은색으로 칠해져 있고, 나머지는 흰색으로 칠해져 있다. 지민이는 이 보드를 잘라서 8×8 크기의 체스판으로 만들려고 한다.

체스판은 검은색과 흰색이 번갈아서 칠해져 있어야 한다. 구체적으로, 각 칸이 검은색과 흰색 중 하나로 색칠되어 있고, 변을 공유하는 두 개의 사각형은 다른 색으로 칠해져 있어야 한다. 따라서 이 정의를 따르면 체스판을 색칠하는 경우는 두 가지뿐이다. 하나는 맨 왼쪽 위 칸이 흰색인 경우, 하나는 검은색인 경우이다.

보드가 체스판처럼 칠해져 있다는 보장이 없어서, 지민이는 8×8 크기의 체스판으로 잘라낸 후에 몇 개의 정사각형을 다시 칠해야겠다고 생각했다. 당연히 8*8 크기는 아무데서나 골라도 된다. 지민이가 다시 칠해야 하는 정사각형의 최소 개수를 구하는 프로그램을 작성하시오.",
  "id": "1018",
  "initialCode": "function solution(input: string[]) {
    // ...
}

solution((global?.process?.env?.arguments || require("fs").readFileSync("/dev/stdin").toString()).trim().split("\\n"));",
  "input": [
    "8 8
WBWBWBWB
BWBWBWBW
WBWBWBWB
BWBBBWBW
WBWBWBWB
BWBWBWBW
WBWBWBWB
BWBWBWBW",
    "10 13
BBBBBBBBWBWBW
BBBBBBBBBWBWB
BBBBBBBBWBWBW
BBBBBBBBBWBWB
BBBBBBBBWBWBW
BBBBBBBBBWBWB
BBBBBBBBWBWBW
BBBBBBBBBWBWB
WWWWWWWWWWBWB
WWWWWWWWWWBWB",
    "8 8
BWBWBWBW
WBWBWBWB
BWBWBWBW
WBWBWBWB
BWBWBWBW
WBWBWBWB
BWBWBWBW
WBWBWBWB",
    "9 23
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBB
BBBBBBBBBBBBBBBBBBBBBBW",
    "10 10
BBBBBBBBBB
BBWBWBWBWB
BWBWBWBWBB
BBWBWBWBWB
BWBWBWBWBB
BBWBWBWBWB
BWBWBWBWBB
BBWBWBWBWB
BWBWBWBWBB
BBBBBBBBBB",
    "8 8
WBWBWBWB
BWBWBWBW
WBWBWBWB
BWBBBWBW
WBWBWBWB
BWBWBWBW
WBWBWWWB
BWBWBWBW",
    "11 12
BWWBWWBWWBWW
BWWBWBBWWBWW
WBWWBWBBWWBW
BWWBWBBWWBWW
WBWWBWBBWWBW
BWWBWBBWWBWW
WBWWBWBBWWBW
BWWBWBWWWBWW
WBWWBWBBWWBW
BWWBWBBWWBWW
WBWWBWBBWWBW",
  ],
  "inputDescription": "첫째 줄에 N과 M이 주어진다. N과 M은 8보다 크거나 같고, 50보다 작거나 같은 자연수이다. 둘째 줄부터 N개의 줄에는 보드의 각 행의 상태가 주어진다. B는 검은색이며, W는 흰색이다.",
  "output": [
    "1",
    "12",
    "0",
    "31",
    "0",
    "2",
    "15",
  ],
  "outputDescription": "첫째 줄에 지민이가 다시 칠해야 하는 정사각형 개수의 최솟값을 출력한다.",
  "title": "체스판 다시 칠하기",
}
`;
